Source/Core/Core/IOS/USB/Bluetooth/LibUSBBluetoothAdapter.cpp

```diff
@@ -0,0 +1,574 @@
+  {
+    DEBUG_LOG_FMT(IOS_WIIMOTE, "Submitting queued HCI command.");
+
+    PushHCICommand(m_pending_hci_transfers.front());
+    m_pending_hci_transfers.pop();
+  }
+
+  CleanCompletedTransfers();
+}
+
+auto LibUSBBluetoothAdapter::ReceiveHCIEvent() -> BufferType
+{
+  if (m_hci_event_queue.Empty())
+  {
+    Update();
+    return {};
+  }
+
+  auto buffer = std::move(m_hci_event_queue.Front());
+  m_hci_event_queue.Pop();
+
+  const auto event = buffer[0];
+  if (event == HCI_EVENT_COMMAND_COMPL)
+  {
+    AcknowledgeCommand<hci_command_compl_ep>(buffer);
+  }
+  else if (event == HCI_EVENT_COMMAND_STATUS)
+  {
+    AcknowledgeCommand<hci_command_status_ep>(buffer);
+  }
+
+  Update();
+  return buffer;
+}
+
+auto LibUSBBluetoothAdapter::ReceiveACLData() -> BufferType
+{
+  BufferType buffer;
+  m_acl_data_queue.Pop(buffer);
+  return buffer;
+}
+
+bool LibUSBBluetoothAdapter::IsControllerReadyForCommand() const
+{
+  return m_num_hci_command_packets > m_unacknowledged_commands.size();
+}
+
+template <typename EventType>
+void LibUSBBluetoothAdapter::AcknowledgeCommand(std::span<const u8> buffer)
+{
+  if (buffer.size() < sizeof(hci_event_hdr_t) + sizeof(EventType)) [[unlikely]]
+  {
+    WARN_LOG_FMT(IOS_WIIMOTE, "Undersized HCI event");
+    return;
+  }
+
+  EventType ev;
+  std::memcpy(&ev, buffer.data() + sizeof(hci_event_hdr_t), sizeof(ev));
+
+  const auto it =
+      std::ranges::find(m_unacknowledged_commands, ev.opcode, &OutstandingCommand::opcode);
+  if (it != m_unacknowledged_commands.end())
+  {
+    DEBUG_LOG_FMT(IOS_WIIMOTE, "HCI command acknowledged: 0x{:04x}", ev.opcode);
+    m_unacknowledged_commands.erase(it);
+  }
+  else if (ev.opcode != 0x0000)
+  {
+    WARN_LOG_FMT(IOS_WIIMOTE, "Unexpected opcode acknowledgement: 0x{:04x}", ev.opcode);
+  }
+
+  m_num_hci_command_packets = ev.num_cmd_pkts;
+}
+
+libusb_transfer* LibUSBBluetoothAdapter::AllocateTransfer(std::size_t buffer_size)
+{
+  auto& item =
+      *m_transfer_buffers.emplace(libusb_alloc_transfer(0), Common::UniqueBuffer<u8>{buffer_size})
+           .first;
+
+  auto* const transfer = item.first;
+  transfer->buffer = item.second.get();
+
+  return transfer;
+}
+
+void LibUSBBluetoothAdapter::FreeTransfer(libusb_transfer* transfer)
+{
+  libusb_free_transfer(transfer);
+  m_transfer_buffers.erase(transfer);
+}
+
+void LibUSBBluetoothAdapter::CleanCompletedTransfers()
+{
+  libusb_transfer* transfer{};
+  while (m_completed_transfers.Pop(transfer))
+    FreeTransfer(transfer);
+}
+
+void LibUSBBluetoothAdapter::PushHCICommand(TimedTransfer transfer)
+{
+  hci_cmd_hdr_t cmd;
+  std::memcpy(&cmd, libusb_control_transfer_get_data(transfer.transfer), sizeof(cmd));
+
+  // Push the time forward so our timeouts work properly if command was queued and delayed.
+  const auto submit_time = std::max(transfer.target_time, Clock::now());
+
+  m_unacknowledged_commands.emplace_back(OutstandingCommand{cmd.opcode, submit_time});
+
+  --m_num_hci_command_packets;
```

Comment:
Maybe a dumb question but if this is 0, should we early out?

Reply:
This function should not be called when `m_num_hci_command_packets` is 0. There are checks before every call. I can add a comment stating that.